synapse/handlers/message.py

```diff
@@ -1461,6 +1462,21 @@ async def persist_and_notify_client_event(
             if prev_state_ids:
                 raise AuthError(403, "Changing the room create event is forbidden")
 
+        if room_version_obj.msc2716_historical and event.type == EventTypes.MSC2716_INSERTION:
+            next_batch_id = event.content.get(EventContentFields.MSC2716_NEXT_BATCH_ID)
+            conflicting_insertion_event_id = await self.store.get_insertion_event_by_batch_id(event.room_id, next_batch_id)
+            if conflicting_insertion_event_id is not None:
+                # The current insertion event that we're processing is invalid
+                # because an insertion event already exists in the room with the
+                # same next_batch_id. We can't allow multiple because the batch
+                # pointing will get weird, e.g. we can't determine which insertion
+                # event the batch event is pointing to.
+                raise SynapseError(
+                    HTTPStatus.BAD_REQUEST,
+                    "Another insertion event already exists with the same next_batch_id",
+                    errcode=Codes.INVALID_PARAM,
+                )
```

Comment:
When we receive these events over federation, should we ignore duplicate insertion events? In a previous iteration of this PR, I ignored duplicate conflicting insertion events that came over federation but really wanted to reject/soft_fail instead (felt more right).

Maybe it's better to just let them do what they want and we can just stop this from the `/send` touch point as a warning stop-gap.